### DISCUSSION TITLE
How do i build a production image?

### DISCUSSION THREAD

**Discussion by Kiskadee-dev** [chatwoot/chatwoot#9577] [Self hosted] [answered]
I must build the project for an architecture that's not officially suported and thereby doesn't have a docker image, but since it's ruby and node it could work as is, so i'm searching what's the official way to build for production but can't find it anywhere besides what's commented in the dockerfile.

**Comment by MarcioLacerda** [comment]
Estou com um problema similar. Estou tentando construir um ambiente com uma maquina arm da OCI, mas só encontrei o procedimento para construir a imagem de desenvolvimento.
https://www.chatwoot.com/docs/contributing-guide/environment-setup/docker/
Queria saber como construir da imagem docker para uma versão específica, se possível.

**Comment by juliancesar** [comment, marked as answer]
To build own image follow this steps:

Git clone from repo:
```
git clone https://github.com/chatwoot/chatwoot.git
git checkout tags/v3.9.0
```

Remove EE features:
```
rm -rf spec/enterprise
rm -rf enterprise
echo -en '\nENV CW_EDITION="ce"' >> docker/Dockerfile
```

Build image (from root folder):
```
docker buildx build --platform linux/amd64 -t USER/chatwoot:v3.9.0 -f ./docker/Dockerfile .
```

> The build process may take 10 minutes!

> This command works on MacOS (M1), maybe you need adapt this, but the --platform is the key of build a image for amd64

Find image ID from builded version:
```
docker image ls | grep chatwoot
```

Tag locally:
```
docker tag ID_IMAGE USER/chatwoot:v3.9.0
```

Send to docker hub:
```
docker push USER/chatwoot:v3.9.0
```

To make this process the docker hub user is required, and when done the image will be available on something like https://hub.docker.com/r/USER/chatwoot.

**Reply by MarcioLacerda** [reply]
Funcionou lindamente. Muito, mas muito obrigado mesmo.
Consegui inclusive gerar uma imagem do 3.7 que precisava pra testar uma migração. Só precisei alterar os comandos abaixo pra rodar na VM ARM da OCI:

`git checkout tags/v3.7.0`
Para voltar pra versão 3.7.0

`docker buildx build --platform linux/arm64 -t marciolacerda/chatwoot:v3.7.0 -f ./docker/Dockerfile .`
Para definir a plataforma como linux/arm64

Não enviei para o docker hub porque ainda estou aprendendo como se faz tudo certo, mas muito brevemente farei isso.

**Reply by FRizzieri** [reply]
Excelente dica!!!
Tenho chatwoot v3.10.2 rodando em uma instancia ARM da OCI em docker.

**Reply by MarcioLacerda** [reply]
Ainda estou aprendendo, mas gerei uma 3.13.0-ce e subi para o dockerhub.
lacinf/chatwoot:v3.13.0-ce

Estou sem tempo pra testar a 3.14.1, mas vou gerar e subir uma versão pra arm até domingo.

**Reply by juliancesar** [reply]
Hoje montei um comando novo que faz quase tudo em uma linha (build em 2 plataformas, tag e push para o docker hub):

```
docker buildx build --no-cache --push --platform linux/amd64,linux/arm64 --tag USER/chatwoot:v3.11.1 -f ./docker/Dockerfile .
```

Outra coisa que vou fazer em breve é testar para ARM, as máquinas na AWS são bem mais baratas!

@FRizzieri, vc usou ARM em produção e tudo deu certo?

**Reply by Harukisatoh** [reply]
Estou tendo esse problema durante o build, alguém saberia alguma solução?

```
> [pre-builder 14/15] RUN git rev-parse HEAD > /app/.git_sha:
0.237 fatal: not a git repository: /app/../.git/modules/chatwoot
------
Dockerfile:78
--------------------
  76 |     
  77 |     # Generate .git_sha file with current commit hash
  78 | >>> RUN git rev-parse HEAD > /app/.git_sha
  79 |     
  80 |     # Remove unnecessary files
--------------------
ERROR: failed to solve: process "/bin/sh -c git rev-parse HEAD > /app/.git_sha" did not complete successfully: exit code: 128
```

**Reply by Harukisatoh** [reply]
> Estou tendo esse problema durante o build, alguém saberia alguma solução?
> 
> ```
> > [pre-builder 14/15] RUN git rev-parse HEAD > /app/.git_sha:
> 0.237 fatal: not a git repository: /app/../.git/modules/chatwoot
> ------
> Dockerfile:78
> --------------------
>   76 |     
>   77 |     # Generate .git_sha file with current commit hash
>   78 | >>> RUN git rev-parse HEAD > /app/.git_sha
>   79 |     
>   80 |     # Remove unnecessary files
> --------------------
> ERROR: failed to solve: process "/bin/sh -c git rev-parse HEAD > /app/.git_sha" did not complete successfully: exit code: 128
> ```

Se você está enfrentando o mesmo problema que eu, descobri o motivo e aqui está a solução!

Esse erro ocorre porque, por algum motivo, o Docker não consegue copiar a pasta `.git` para dentro do container. Sem essa pasta, o comando `RUN git rev-parse HEAD > /app/.git_sha` falha, já que a pasta `/app` dentro do container não será reconhecida como um repositório Git.

No meu caso, o problema acontecia porque eu estava tentando fazer o build dentro de um submódulo do Git. (Se quiser saber mais sobre submódulos, confira: https://git-scm.com/book/en/v2/Git-Tools-Submodules). Como um submódulo não possui a pasta `.git` diretamente (ela é referenciada de outra forma), o Docker não conseguia copiá-la para o container.

Se você apenas clonou o repositório do Chatwoot e tentou fazer o build, esse erro provavelmente não acontecerá com você.

**Comment by FRizzieri** [comment]
@juliancesar, tenho utilizado em produção sem problemas, mas com uma
observação:

Já compilei o meu "build" Arm64 direto na plataforma Arm64.
O Oracle cloud usa processadores Ampére, já a AWS usa o graviton (na AWS eu
não testei)

Em qua, 6 de nov de 2024 18:39, Julian Cesar ***@***.***>
escreveu:
